### PR TITLE
QI Transfers

### PR DESCRIPTION
**Pull request by 0xtaf**
# QI Transfers

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [x] New Agent
- [ ] Fix an agent error
- [ ] Update an agent logic/test/documentation
- [ ] Code style update (formatting, renaming)
- [ ] Other (please describe):

## New Agent PR readiness

- [x] `npm install` command works (i.e. package-lock is updated properly)
- [x] The `name` and `description` fields of `package.json` describe the agent properly
- [x] All the libraries listed in `package.json` are being used
- [x] All the development libraries are in `devDependencies` section of `package.json`
- [x] The agent README is updated
- [x] The agent README sufficiently describes what the agent does
- [x] All the alerts that the agent emits are listed in the README
- [x] The txn/block examples listed in the README generate the expected findings
- [x] Code is easy to follow and/or properly commented
- [x] Agent contains tests
- [x] All tests pass when running `npm test`
- [x] Tests are not making real network calls (i.e. network interaction is properly mocked)
- [x] Tests cover all the paths (i.e. negative & positive test cases)
- [x] When the tests complete, no warning/errors are shown
- [x] `npm start` works without errors when executed using a jsonRpcUrl for each of the networks specified in the README
- [x] The code does what the README describes
- [x] Network calls are properly being cached (if duplicated calls can occur)
- [x] A performance review has been done (e.g. `Promise.all` is used whenever possible)
- [x] The findings fields are filled out correctly
- [x] Findings `protocol` is specified (if the agent is for a specific one)
- [x] `metadata` of the findings contains all the information needed.
- [x] The SDK methods are being used appropriately (e.g. `filterLog`, `filterFunction`, `getEthersProvider`)
- [x] `ABI`s match the contracts used / `args` used from `LogDescriptions` exists in the `ABI`s
- [x] The agent follows Forta recommended [best practices](https://docs.forta.network/en/latest/best-practices/)
- [ ] Agent can recover from failed error calls when they can fail (e.g. errors are being caught on calls that can fail)

### CODE REVIEW

File: Benqi-Agents/Qi-Transfers/src/data.fetcher.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+  constructor(qiTokenAddress: string, provider: providers.Provider) {
+    this.qiTokenAddress = qiTokenAddress;
+    this.provider = provider;
+    this.qiTokenContract = new Contract(qiTokenAddress, benqiInterface, provider);
+  }
+
+  // return the number of tokens held by the account
+  public async getBalance(account: string): Promise<BigNumber> {
+    const balance: BigNumber = await this.qiTokenContract.balanceOf(account, { blockTag: "latest" });
```
Comment (e1Ru1o) [inline Benqi-Agents/Qi-Transfers/src/data.fetcher.ts:17]: why are you fetching from `latest` instead of an specific block?

File: Benqi-Agents/Qi-Transfers/src/agent.spec.ts [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+    resetAllWhenMocks();
+
+    handleTransaction = provideHandleTransaction(mockFetcher as any);
+  });
+
+  it("should return empty finding if transferred amount and account balance are below threshold", async () => {
+    when(mockFetcher.getBalance).calledWith(testAccounts[2]).mockReturnValue(testBalances[2]);
+
+    const log1 = TEST_BENQI_IFACE.encodeEventLog(TEST_BENQI_IFACE.getEvent("Transfer"), [
+      createAddress("aaa"), // address from
```
Comment (e1Ru1o) [inline Benqi-Agents/Qi-Transfers/src/agent.spec.ts:63]: missing `0x` in `0xaaa`. Same applies to all the tests below

File: Benqi-Agents/Qi-Transfers/src/agent.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+      const toAddress = log.args.to;
+      const balance = await fetcher.getBalance(toAddress);
+
+      if (tokenAmount.gt(TRANSFERRED_TOKEN_THRESHOLD)) {
+        findings.push(createTransferFinding(log));
+      }
+
+      if (balance.gt(BALANCE_THRESHOLD)) {
+        findings.push(createLargeBalanceFinding(toAddress, balance));
+      }
```
Comment (e1Ru1o) [inline Benqi-Agents/Qi-Transfers/src/agent.ts:35]: Please pass this thresholds as parameters, in order to isolate the tests and allow them to always pass even if some numbers are changed in utils file. About that I would like to

**Review by e1Ru1o — CHANGES_REQUESTED** [review]
- `npm install` show one high severity --> `npm audit` fix required
- use `Benqi Finance` in protocol property of the findings

File: Benqi-Agents/Qi-Transfers/src/data.fetcher.ts (hunk 1 above)
Comment (0xtaf) [reply Benqi-Agents/Qi-Transfers/src/data.fetcher.ts:17]: because `balanceOf` function in the contract only takes an `account` as a parameter. and I thought whenever it is called, the most updated balance of an account is important. should it be for the specific block in which agent is running on at that time?

Comment (e1Ru1o) [reply Benqi-Agents/Qi-Transfers/src/data.fetcher.ts:17]: `blockTag` let you specify the block from where you want to take the data. That means that you will be calling `balanceOf` in that specific block, so the value returned will be the one the address had in that block instead of the most updated balance. The the answer to the question is yes.

File: Benqi-Agents/Qi-Transfers/src/agent.spec.ts [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+      .setBlock(testBlock);
+    const findings = await handleTransaction(transactionEvent);
+
+    expect(findings).toStrictEqual([
+      createLargeBalanceFinding(testAccounts[0], testBalances[0]),
+      createLargeBalanceFinding(testAccounts[1], testBalances[1]),
+    ]);
+  });
+
+  it("should return findings if transferred amount and account balance are above threshold", async () => {
```
Comment (yasmine9) [inline Benqi-Agents/Qi-Transfers/src/agent.spec.ts:197]:
```suggestion
  it("should return 2 findings if transferred amount and account balance are above threshold", async () => {
```

File: Benqi-Agents/Qi-Transfers/src/agent.spec.ts [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+    protocol: "Benqi Finance",
+    metadata: {
+      from: logDesc.args.from,
+      to: logDesc.args.to,
+      amount: BigNumber.from(logDesc.args.amount).toString(),
+    },
+  });
+};
+
+describe("Benqi Token (QI) Transfer and Balance Tests", () => {
```
Comment (yasmine9) [inline Benqi-Agents/Qi-Transfers/src/agent.spec.ts:52]: I recommend adding the test case for empty transactions

**Review by yasmine9 — CHANGES_REQUESTED** [review]

File: Benqi-Agents/Qi-Transfers/README.md [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    - from: Source address
+    - to: Destination address
+    - amount: The amount of transferred tokens
+
+- BENQI-3-2
+  - Fired when balance of the destination account of a transfer is more than threshold percentage of the total supply
+  - Severity is always set to "Info"
+  - Type is always set to "Info"
+  - Metadata includes:
+    - account: The account that gets the tokens
```
[the remark this answers is not in the document]
Comment (0xtaf) [reply Benqi-Agents/Qi-Transfers/README.md:40]: here I mean that it's the account in the `toAddress` because the sender and the receiver and anyone else might hold the tokens but the agent only checks the balance of the receiver address.

File: Benqi-Agents/Qi-Transfers/src/utils.ts [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+export const benqiInterface = new ethers.utils.Interface(BALANCE_ABI);
+
+const isTestnet: boolean = false;
+
+export const QI_TOKEN_CONTRACT = isTestnet
+  ? "0x749590F04a3aa91B53E03124FDa4Ec0391D4Dd4E" // AVAX Testnet QI Token Contract address
+  : "0x8729438EB15e2C8B576fCc6AeCdA6A148776C0F5"; // AVAX Mainnet QI Token Contract address
+
+const QI_DECIMALS: number = 18;
+const TOTAL_SUPPLY: BigNumber = BigNumber.from("7200000000").mul(`${10 ** QI_DECIMALS}`);
```
[the remark this answers is not in the document]
Comment (0xtaf) [reply Benqi-Agents/Qi-Transfers/src/utils.ts:15]: I also thought doing it that way. But then I decided that since it's a specific number that will never change, I wanted to avoid any possible problems while fetching the total supply data. still not very sure which one makes more sense.

File: Benqi-Agents/Qi-Transfers/src/data.fetcher.ts [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+    this.provider = provider;
+    this.qiTokenContract = new Contract(qiTokenAddress, benqiInterface, provider);
+  }
+
+  // return the number of tokens held by the account
+  public async getBalance(account: string, blockNumber: number): Promise<BigNumber> {
+    const balance: BigNumber = await this.qiTokenContract.balanceOf(account, { blockTag: blockNumber });
+    return balance;
+  }
+}
```
[the remark this answers is not in the document]
Comment (0xtaf) [reply Benqi-Agents/Qi-Transfers/src/data.fetcher.ts:20]: just lack of knowledge because I've never used it before. I'll learn how to do it. thanks for the suggestion

File: Benqi-Agents/Qi-Transfers/src/agent.spec.ts [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+    name: "QI Token Transfer",
+    description: "Large amount of QI token transfer is detected",
+    alertId: "BENQI-3-1",
+    severity: FindingSeverity.Info,
+    type: FindingType.Info,
+    protocol: "Benqi Finance",
+    metadata: {
+      from: logDesc.args.from,
+      to: logDesc.args.to,
+      amount: BigNumber.from(logDesc.args.amount).toString(),
```
[the remark this answers is not in the document]
Comment (0xtaf) [reply Benqi-Agents/Qi-Transfers/src/agent.spec.ts:47]: you're right. fixing it

File: Benqi-Agents/Qi-Transfers/src/utils.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (e1Ru1o) [reply Benqi-Agents/Qi-Transfers/src/utils.ts:15]: You did well, that value is constant in the contract.

File: Benqi-Agents/Qi-Transfers/jest-when.d.ts [hunk 10]
```diff
@@ -0,0 +1 @@
+declare module "jest-when";
```
Comment (e1Ru1o) [inline Benqi-Agents/Qi-Transfers/jest-when.d.ts:1]: What is this? And why is needed?

File: Benqi-Agents/Qi-Transfers/src/data.fetcher.ts [hunk 11]
```diff
@@ -1,20 +1,28 @@
...
   constructor(qiTokenAddress: string, provider: providers.Provider) {
     this.qiTokenAddress = qiTokenAddress;
     this.provider = provider;
     this.qiTokenContract = new Contract(qiTokenAddress, benqiInterface, provider);
+    this.cache = new LRU<string, BigNumber>({ max: 10000 });
   }
 
   // return the number of tokens held by the account
   public async getBalance(account: string, blockNumber: number): Promise<BigNumber> {
+    const key: string = `${account + blockNumber}`;
```
Comment (e1Ru1o) [inline Benqi-Agents/Qi-Transfers/src/data.fetcher.ts:20]:
```suggestion
    const key: string = `${account}-${blockNumber}`;
```
`string + number` doesn't look like a good practice for me. Please consider to apply this suggestion.

File: Benqi-Agents/Qi-Transfers/src/data.fetcher.ts [hunk 12]
```diff
@@ -1,20 +1,28 @@
...
+
+    if (this.cache.get(key) !== undefined) return this.cache.get(key) as BigNumber;
```
Comment (e1Ru1o) [inline Benqi-Agents/Qi-Transfers/src/data.fetcher.ts:22]:
```suggestion
    if (this.cache.has(key)) return this.cache.get(key) as BigNumber;
```
You can use `has` function here

File: Benqi-Agents/Qi-Transfers/package.json [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+    "publish": "forta-agent publish",
+    "push": "forta-agent push",
+    "disable": "forta-agent disable",
+    "enable": "forta-agent enable",
+    "keyfile": "forta-agent keyfile",
+    "test": "jest",
+    "format": "prettier --write \"src/**/*.ts\""
+  },
+  "dependencies": {
+    "@types/lru-cache": "^7.5.0",
```
Comment (e1Ru1o) [inline Benqi-Agents/Qi-Transfers/package.json:23]: Move this to `devDependencies`

File: Benqi-Agents/Qi-Transfers/jest-when.d.ts (hunk 10 above)
Comment (0xtaf) [reply Benqi-Agents/Qi-Transfers/jest-when.d.ts:1]: it was giving me a yellow warning since jest-when has implicitly `any` type although I installed @types/jest-when. so I had to explicitly declare the type. but now I removed it and it doesn't show me the error. I guess it was a glitch in my vscode then.

**Review by e1Ru1o — APPROVED** [review]
